### DISCUSSION TITLE
Validating Syntax of Dynamic Expressions without parameter values

### DISCUSSION THREAD

**Discussion by aerott** [davideicardi/DynamicExpresso#326] [Q&A] [answered]
Hi community,

I have a system that allows creating rules, rule always returns bool.
I define an expression in a Rule, for example, `AGE > 50 && GROWTH <= 180`.
I use Interpreter to evaluate the Rule, but this is a long operation because I must query the database for each parameter value.
 
My parameters are defined like this:
```
public class RuleParameter {
    public string Name { get; set; }
    public ParameterValueKind? ValueKind { get; private set; }
    ...
}
 
public enum ParameterValueKind
{
    String,
    Boolean,
    Integer,
    Decimal,
    DateTime
}
```
 
I then use `DynamicExpresso.Parameter.Create(ruleParameter.Name, value)` to create a parameter and call `interpreter.Eval<bool>(rule.Expression, ruleParameters.ToArray());`
 
I'd like to add validation to my system, so I need to check if the expression that user entered in my form is valid.
From user input I know what parameters will be used and I know the expression.
 
I'm doing this right now:
```
var interpreter = new Interpreter(InterpreterOptions.Default | InterpreterOptions.LambdaExpressions | InterpreterOptions.DefaultCaseInsensitive);
 
var testParameters = new List<Parameter>
{
Parameter.Create("a", ParameterValueKind.Integer, "test"),
Parameter.Create("b", ParameterValueKind.Integer, "test"),
Parameter.Create("c", ParameterValueKind.Integer, "test")
};
 
var testRule = Rule.Create("test", "a + b + c > 10", testParameters);
 
var expParameters = new List<DynamicExpresso.Parameter>();
 
foreach (var parameter in rule.Parameters)
{
switch (parameter.ValueKind)
{
	case ParameterValueKind.Integer:
		expParameters.Add(DynamicExpresso.Parameter.Create(parameter.Name, 1));
		break;
	case ParameterValueKind.String:
		expParameters.Add(DynamicExpresso.Parameter.Create(parameter.Name, "test"));
		break;
	case ParameterValueKind.Boolean:
		expParameters.Add(DynamicExpresso.Parameter.Create(parameter.Name, true));
		break;
	case ParameterValueKind.Decimal:
		expParameters.Add(DynamicExpresso.Parameter.Create(parameter.Name, 1.0));
		break;
	case ParameterValueKind.DateTime:
		expParameters.Add(DynamicExpresso.Parameter.Create(parameter.Name, DateTime.Now));
		break;
	default:
		throw new ArgumentOutOfRangeException();
}
}
var isValid = true;
try
{
interpreter.Parse(rule.Definition, expParameters.ToArray());
}
catch (Exception ex)
{
isValid = false;
}
```
 
is there an easier way?

**Comment by davideicardi** [comment, marked as answer]
You don't have to specify the parameter value in order to parse the expression.
Here an example:
```
var target = new Interpreter();

var parameters = new[] {
	new Parameter("x", typeof(int)),
	new Parameter("y", typeof(int))
};

var myFunc = target.Parse("x + y", parameters);

Assert.AreEqual(30, myFunc.Invoke(23, 7));
Assert.AreEqual(30, myFunc.Invoke(32, -2));
```

If the expression is not valid, it will throw in the `Parse` function.

**Reply by aerott** [reply]
Hi @davideicardi,

Many thanks. It looks like your solution is sufficient for me :)

I got a side question – is there a way to extract information about which parameter is problematic? 

E.g.: expression: `x + y.First()` 
Right now, I get messages like `No applicable method 'First' in 'int' (at index 6),` but that's not necessarily a clear hint for a more complex expression. 

Can I extract information about which parameter it is referring to?

**Reply by davideicardi** [reply]
Unfortunately I don't think you can get much more info. You have the index but the messages are always a bit technical.
